### DISCUSSION TITLE
Annotate all error results with context.

### DISCUSSION THREAD

**Discussion by joeblubaugh** [grafana/grafana#48848] [Alerting]
Grafana 8 alerting lets you define states for query execution errors and no data conditions. You can have them alert, have them be "OK" or have them notify you with the states "Error" and "No Data".

In the error condition mapping code there's a switch:
```go
execErrState := eval.Error
switch alertRule.ExecErrState {
case ngModels.AlertingErrState:
	execErrState = eval.Alerting
case ngModels.ErrorErrState:
	// If the evaluation failed because a query returned an error then
	// update the state with the Datasource UID as a label and the error
	// message as an annotation so other code can use this metadata to
	// add context to alerts
	var queryError expr.QueryError
	if errors.As(a.Error, &queryError) {
		for _, next := range alertRule.Data {
			if next.RefID == queryError.RefID {
				a.Labels["ref_id"] = next.RefID
				a.Labels["datasource_uid"] = next.DatasourceUID
				break
			}
		}
		a.Annotations["Error"] = queryError.Error()
	}
	execErrState = eval.Error
case ngModels.OkErrState:
	a.resultNormal(alertRule, result)
	return
default:
	a.Error = fmt.Errorf("cannot map error to a state because option [%s] is not supported. evaluation error: %w", alertRule.ExecErrState, a.Error)
}
```

In the `ngModels.ErrorErrState` block, the alert state gets additional annotations that explain the error, including where an error in the query happened.

Would it be useful to also have these annotations on alerting state? As a notification receiver, I might very well be interested in these errors, as they could help me root-cause the problem quickly.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
